src/timemachine-heap.c

```diff
@@ -0,0 +1,150 @@
+/* Copyright (C) 2007-2014 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+
+/**
+ * \file
+ *
+ * \author Mat Oldham <mat.oldham@gmail.com>
+ *
+ * Heap related functions for usage within TimeMachine
+ */
+
+#include "suricata-common.h"
+#include "timemachine.h"
+#include "timemachine-heap.h"
+#include "timemachine-packet.h"
+
+TimeMachineHeap* TimeMachineHeapNew(TimeMachineThreadVars* tmtv, TimeMachineHeapConf* conf) {
+
+    TimeMachineHeap *heap = SCMalloc(sizeof(TimeMachineHeap));  
+    if (unlikely(heap == NULL)) {
+        return NULL;     
+    }
+
+    heap->conf = conf;
+    heap->used_pool_count = 0;
+    heap->unused_pool_count = 0;
+    
+    TAILQ_INIT(&heap->used_mem_pools);
+    TAILQ_INIT(&heap->unused_mem_pools);
+    TAILQ_INIT(&heap->unused_packets);    
+    
+    uint32_t expanded = TimeMachineHeapExpand(tmtv, heap, conf->prealloc_count);
+    if (expanded != conf->prealloc_count) {
+        SCLogError(SC_ERR_MEM_ALLOC, "Error: Could not allocate initial memory for "
+                   "time machine, expanded heap by: %" PRIu32 ", requested: %" PRIu32 
+                   " current Memory: %" PRIu64 " max Memory: %" PRIu64, expanded, 
+                   conf->prealloc_count, tmtv->current_memory, timemachine_config.max_memory);
+        exit(EXIT_FAILURE);                         
```

Comment:
TimeMachineNew isn't actually called during packet runtime (can add that as documentation).  The initial TimeMachine heaps are created during thread vars allocation.  I think we would rather fail out if we don't have the memory for the initial TimeMachine heaps as part of the thread allocation.  Subsequent expansions would be in the packet runtime and therefore would just not expand the heap at that point.  Still an issue?